#!/bin/bash
#
#   +-----------------------+
#   |  USE:                 |
#   |    - Medium queue     |
#   |    - 2 CPU + 4 Gb     |
#   +-----------------------+
#
#  Replace the USER name in this script with your username and
#  call your project whatever you want
#
#  This script must be made executable like this
#    chmod +x my_script
#
#  Submit this script to the queue with a command like this
#    run_script my_script.sh
#
#  My preferred setup before running:
#    -- script to be run in /home/projectdir/scripts
#    -- project directory (of same name as script) in /home/
#    -- /input/ and /output/ subdirs within project dir

# -----------------------------------------------------------------------------
# Set variables for this step
# -----------------------------------------------------------------------------

STEP=02_read_sim_and_qc
PREV_STEP=01_run_slim
SCRIPT=02a_run_art.sh

# -----------------------------------------------------------------------------
# Load variables and functions from settings file
# -----------------------------------------------------------------------------

source /home/aubkbk001/roh_param_project/scripts/99_includes/init_script_vars.sh

# -----------------------------------------------------------------------------
# Load modules
# -----------------------------------------------------------------------------

module load anaconda/3-2020.11

# -----------------------------------------------------------------------------
# Run read simulation
#
# INPUT_DIR, SLIM_OUT_DIR, FASTA_OUT_DIR, OUTPUT_DIR, and SAMPLE_ID_LIST are
# set in  init_script_vars.sh
# -----------------------------------------------------------------------------

# Starting with 50X coverage for now (25X per haploid genome, i.e., FASTA file)
# Will downsample unfiltered BAM files generated by BWA.
# 150-bp paired-end reads (--len and --paired parameters).
# Using the HiSeq 2500 error model (--seqSys parameter).

while read -a line; do

    for a in 1 2; do

        start_logging "ART Read - ${line[0]}_$a.fasta"

        art_illumina \
            --seqSys HS25 \
            -i ${FASTA_OUT_DIR}/${line[0]}_$a.fasta \
            --paired \
            -na \
            --len 150 \
            --fcov 25 \
            -m 500 \
            -s 75 \
            -o ${OUTPUT_DIR}/${line[0]}_$a

        stop_logging

    done

done <${SAMPLE_ID_LIST}

# mail -s 'ART run finished' kirkseykb1@appstate.edu <<<'SLiM run finished'

# -----------------------------------------------------------------------------
# Copy output files to user's home directory.
# -----------------------------------------------------------------------------

source /home/aubkbk001/roh_param_project/scripts/99_includes/backup_output.sh
