#!/bin/bash
#
#   +-----------------------+
#   |  USE:                 |
#   |    - Medium queue     |
#   |    - 2 CPU + 4 Gb     |
#   +-----------------------+
#
#  Replace the USER name in this script with your username and
#  call your project whatever you want
#
#  This script must be made executable like this
#    chmod +x my_script
#
#  Submit this script to the queue with a command like this
#    run_script my_script.sh
#
#  My preferred setup before running:
#    -- script to be run in /home/projectdir/scripts
#    -- project directory (of same name as script) in /home/
#    -- /input/ and /output/ subdirs within project dir

##  Set username
USER=aubkbk001
#foo
## Set project name
PROJ=02_read_sim_and_qc

## Create a directory on /scratch
mkdir /scratch/${USER}_${PROJ}/

## Set permissions for directory
chmod 700 /scratch/${USER}_${PROJ}/

## cd into working scratch directory
cd /scratch/${USER}_${PROJ}/

## --------------------------------
## Load modules
module load anaconda/3-2020.11


## --------------------------------
## Run read simulation
## (running just 'art_illumina' brings up usage, parameters)
INDIR=/scratch/aubkbk001_01_slim/

## Starting with 50X coverage for now (25X per haploid genome, i.e., FASTA file); will 
## downsample unfiltered BAM files generated by BWA.
## 150-bp paired-end reads (--len and --paired parameters).
## Using the HiSeq 2500 error model (--seqSys parameter).
mkdir sim_fastq_files
chmod 700 sim_fastq_files
cd sim_fastq_files

while read -a line
	do
	for a in 1 2
		do
        echo ${INDIR}sample_fasta_files_m5e-07_r1e-8_p500/${line[0]}_$a.fasta
        echo ${line[0]}_$a
		art_illumina \
		--seqSys HS25 \
		-i ${INDIR}sample_fasta_files_m5e-07_r1e-8_p500/${line[0]}_$a.fasta \
		--paired \
		-na \
		--len 150 \
		--fcov 25 \
		-m 500 \
		-s 75 \
		-o ${line[0]}_$a
		done
	done < ${INDIR}sample_id_list_m5e-07_r1e-8_p500.txt

mail -s 'ART run finished' kirkseykb1@appstate.edu <<< 'SLiM run finished'

## Starting with 50X coverage for now (25X per haploid genome, i.e., FASTA file); will 
## downsample unfiltered BAM files generated by BWA.
## 150-bp paired-end reads (--len and --paired parameters).
## Using the HiSeq 2500 error model (--seqSys parameter).
# mkdir sim_fastq_files/
# cd sim_fastq_files/
# while read -a line
# 	do
# 	for a in 1 2
# 		do
# 		art_illumina \
# 		--seqSys HS25 \
# 		-i ${INDIR}sample_fasta_files_m5e-07_r1e-8_p500/${line[0]}_a.fasta \
# 		--paired \
# 		-na \
# 		--len 150 \
# 		--fcov 25 \
# 		-m 500 \
# 		-s 75 \
# 		-o $i_a
# 		done
# 	done < ${INDIR}sample_id_list_m5e-07_r1e-8_p500.txt

# mail -s 'ART run finished' kirkseykb1@appstate.edu <<< 'SLiM run finished'